### PR TITLE
Migrate Tlc1543 device to nanoFramework

### PR DESCRIPTION
**Pull request by usmanatron**
## Description
- Update the Tlc1543 device to use the nanoFrameowkr
- I had a look at another similar PR to understand the scope of the changes (nanoframework/nanoFramework.IoT.Device!192)
- This involved updating a lot of the packages and generally finding where things now live 

## Motivation and Context
- Resolves nanoFramework/Home#865

## How Has This Been Tested?
This has not been tested because I have no idea how to do so (I don't have one of these!).  If there's anything you recommend around here please let me know.  The changes themselves should be self-contained though.

## Questions
1. There's one specific area I'm a bit unsure about, around the Spi constructor.  The Samples project sets a load of specific pin variables, which no longer are available to set.  I don't have the context to know how important these are.  Any feedback would be appreciated!
2. Should I delete the old Tlc1543 device files from `src/devices_generated`?

## Types of changes
- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [x] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (work on Unit Tests, has no impact on code or features)
- [x] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/nanoframework/nanoFramework.IoT.Device?pullRequest=211) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/nanoframework/nanoFramework.IoT.Device?pullRequest=211) before we can accept your contribution.<br/><br/>:x: usmanatron [sign now](https://cla.dotnetfoundation.org/nanoframework/nanoFramework.IoT.Device?pullRequest=211)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/nanoframework/nanoFramework.IoT.Device?pullRequest=211) it.</sub>

**Comment by usmanatron** [comment]
Note: I forgot to update the nuspec file!  I'll sort that (and the build)

File: devices/Tlc1543/Tlc1543.nfproj [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    <None Include="packages.config" />
+    <None Include="README.md" />
+  </ItemGroup>
+  <ItemGroup>
+    <Compile Include="Properties\AssemblyInfo.cs" />
+    <Compile Include="*.cs" />
+    <None Include="*.md" />
+  </ItemGroup>
+  <ItemGroup>
+    <Reference Include="mscorlib">
```
Comment (Ellerbach) [inline devices/Tlc1543/Tlc1543.nfproj:32]: Here, it should be the full qualified elements like here: https://github.com/nanoframework/nanoFramework.IoT.Device/blob/a3b31ab4efc1dfeed83610a56908c988ce718466/devices/AD5328/AD5328.nfproj#L23

And it does apply to all of them.

File: devices/Tlc1543/Tlc1543.nuspec [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    <developmentDependency>false</developmentDependency>
+    <projectUrl>https://github.com/nanoframework/nanoFramework.IoT.Device</projectUrl>
+    <icon>images\nf-logo.png</icon>
+    <repository type="git" url="https://github.com/nanoframework/nanoFramework.IoT.Device" commit="$commit$" />
+    <copyright>Copyright (c) .NET Foundation and Contributors</copyright>
+    <description>This package includes the .NET IoT Core binding Iot.Device.Tlc1543 for .NET nanoFramework C# projects.</description>
+    <summary>Iot.Device.Tlc1543 assembly for .NET nanoFramework C# projects</summary>
+    <tags>nanoFramework C# csharp netmf netnf Iot.Device.Tlc1543</tags>
+    <dependencies>
+      <dependency id="nanoFramework.CoreLibrary" version="1.10.4-preview.11" />
```
Comment (Ellerbach) [inline devices/Tlc1543/Tlc1543.nuspec]: You need to add all the dependencies here. So all the nuget you are using with the correct version as well. It's not done automatically.

File: devices/Tlc1543/samples/Tlc1543.Sample.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    /// Samples for Tlc1543
+    /// </summary>
+    public class Program
+    {
+        /// <summary>
+        /// Main entry point
+        /// </summary>
+        public static void Main()
+        {
+            SpiDevice spi = SpiDevice.Create(new SpiConnectionSettings(-1) { DataBitLength = Tlc1543.SpiDataBitLength, ChipSelectLine = 5,  });
```
Comment (Ellerbach) [inline devices/Tlc1543/samples/Tlc1543.Sample.cs:20]: Please add a message like this one on SPI setup for devices like ESP32:
https://github.com/nanoframework/nanoFramework.IoT.Device/blob/a3b31ab4efc1dfeed83610a56908c988ce718466/devices/Pn532/samples/Program.cs#L30

In short, it's about adding this:
```csharp
//////////////////////////////////////////////////////////////////////
// when connecting to an ESP32 device, need to configure the SPI GPIOs
// used for the bus
//Configuration.SetPinFunction(21, DeviceFunction.SPI1_MOSI);
//Configuration.SetPinFunction(22, DeviceFunction.SPI1_MISO);
//Configuration.SetPinFunction(23, DeviceFunction.SPI1_CLOCK);
```

Comment (Ellerbach) [reply devices/Tlc1543/samples/Tlc1543.Sample.cs:20]: Also, in this case, -1 won't work, should be a valid one

File: devices/Tlc1543/samples/Tlc1543.Samples.nfproj [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    <Compile Include="Properties\AssemblyInfo.cs" />
+    <Compile Include="*.cs" />
+    <None Include="*.md" />
+    <None Include="packages.config" />
+  </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\Tlc1543.nfproj" />
+  </ItemGroup>
+  <ItemGroup>
+    <Reference Include="mscorlib">
```
Comment (Ellerbach) [inline devices/Tlc1543/samples/Tlc1543.Samples.nfproj:32]: same here as for the previous one, should be the full qualified ones.
Reason: for bump and build in the pipeline

**Review by Ellerbach — COMMENTED** [review]
Thanks a lot for the effort! Looks great on the code part! Some adjustments needed on the project csproj and nugets! For the rest nice work!
Did you manage to test it on a real device?

**Comment by Ellerbach** [comment]
> Note: I forgot to update the nuspec file! I'll sort that (and the build)

And spotted that in the review ;-)

File: devices/Tlc1543/Tlc1543.nuspec [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    </dependencies>
+  </metadata>
+  <files>
+    <file src="bin\Release\Iot.Device.Tlc1543.dll" target="lib\Iot.Device.Tlc1543.dll" />
+    <file src="bin\Release\Iot.Device.Tlc1543.pdb" target="lib\Iot.Device.Tlc1543.pdb" />
+    <file src="bin\Release\Iot.Device.Tlc1543.pdbx" target="lib\Iot.Device.Tlc1543.pdbx" />
+    <file src="bin\Release\Iot.Device.Tlc1543.pe" target="lib\Iot.Device.Tlc1543.pe" />
+    <file src="bin\Release\Iot.Device.Tlc1543.xml" target="lib\Iot.Device.Tlc1543.xml" />
+    <file src="README.md" target="docs\" />
+    <file src="..\assets\nf-logo.png" target="images" />
```
Comment (Ellerbach) [inline devices/Tlc1543/Tlc1543.nuspec:38]:
```suggestion
    <file src="..\..\assets\nf-logo.png" target="images" />
```

File: devices/Tlc1543/Tlc1543.nuspec [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    <file src="..\LICENSE.md" target="" />
```
Comment (Ellerbach) [inline devices/Tlc1543/Tlc1543.nuspec:39]:
```suggestion
    <file src="..\..\LICENSE.md" target="" />
```

**Comment by Ellerbach** [comment]
> This has not been tested because I have no idea how to do so (I don't have one of these!). If there's anything you recommend around here please let me know. The changes themselves should be self-contained though.

Well, there is not much you can do for testing... You need to have the device. But on the code side, with the adjustment, it seems ok. One basic test you can do is: deploy the sample on the device. If anything bad (like missing generics in the code), you'll see it right away. The code should work up to the device creation, including SPI (you may have to properly setup the pins in case of ESP32) and that's going to be about it.

**Comment by usmanatron** [comment]
Thanks @Ellerbach - all of your comments make sense to me (and answer my question about the pin numbers!).  I'll make those changes

File: devices/Tlc1543/Tlc1543.nfproj [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    </Reference>
+    <Reference Include="System.Diagnostics.Stopwatch, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null">
+      <HintPath>packages\nanoFramework.System.Diagnostics.Stopwatch.1.0.259\lib\System.Diagnostics.Stopwatch.dll</HintPath>
+    </Reference>
+    <Reference Include="System.Math, Version=1.4.3.3, Culture=neutral, PublicKeyToken=c07d481e9758c731">
+      <HintPath>packages\nanoFramework.System.Math.1.4.3\lib\System.Math.dll</HintPath>
+    </Reference>
+  </ItemGroup>
+  <Import Project="$(NanoFrameworkProjectSystemPath)NFProjectSystem.CSharp.targets" Condition="Exists('$(NanoFrameworkProjectSystemPath)NFProjectSystem.CSharp.targets')" />
+  <Import Project="..\..\src\System.Runtime.CompilerService\System.Runtime.CompilerService.projitems" Label="Shared" />
```
Comment (Ellerbach) [inline devices/Tlc1543/Tlc1543.nfproj:53]: I don't think this project is needed. If you remove it, does the project still build?

File: devices/Tlc1543/Tlc1543.nfproj [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    <Reference Include="nanoFramework.Runtime.Events, Version=1.9.2.3, Culture=neutral, PublicKeyToken=c07d481e9758c731">
+      <HintPath>packages\nanoFramework.Runtime.Events.1.9.2\lib\nanoFramework.Runtime.Events.dll</HintPath>
+    </Reference>
+    <Reference Include="System.Device.Gpio, Version=1.0.2.3, Culture=neutral, PublicKeyToken=c07d481e9758c731">
+      <HintPath>packages\nanoFramework.System.Device.Gpio.1.0.2\lib\System.Device.Gpio.dll</HintPath>
+    </Reference>
+    <Reference Include="System.Device.Spi, Version=1.0.2.3, Culture=neutral, PublicKeyToken=c07d481e9758c731">
+      <HintPath>packages\nanoFramework.System.Device.Spi.1.0.2\lib\System.Device.Spi.dll</HintPath>
+    </Reference>
+    <Reference Include="System.Diagnostics.Stopwatch, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null">
```
Comment (Ellerbach) [inline devices/Tlc1543/Tlc1543.nfproj:45]: are you sure there is no public token in this one? (not sure reason why I'm asking)

File: devices/Tlc1543/samples/version.json [hunk 9]
```diff
@@ -0,0 +1,19 @@
+{
```
Comment (Ellerbach) [inline devices/Tlc1543/samples/version.json:1]: you can delete this file. Not needed in the sample folder

**Review by Ellerbach — COMMENTED** [review]
Looks very good, very little thing to check!

File: devices/Tlc1543/README.md [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+### Simple Line finding algorithm with ADC usage
+
+This sample shows how to calculate position of line under matrix of 5 IR sensors with the help of TLC1543 ADC. In this case it's AlphaBot2 Pi from WaveShare with ITR-20001 IR sensors.
+
+### Initialization
+
+```csharp
+SpiDevice spi = SpiDevice.Create(
+    new SpiConnectionSettings(-1) { DataBitLength = Tlc1543.SpiDataBitLength, ChipSelectLine = 5 });
```
Comment (Ellerbach) [inline devices/Tlc1543/README.md:19]: please adjust the setting here as well ;-)

**Comment by josesimoes** [comment]
Hey @usmanatron! 😄 
Any chance you can fix these tiny bits so this can be merged? It's sooo close... 😉

**Comment by usmanatron** [comment]
Hey @josesimoes.  Apologies for the silence, I tried to have a screen break over the holidays!
Definitely keen to finish this.  I should have time over the next day or two to get it done.

File: devices/Tlc1543/Tlc1543.nfproj (hunk 8 above)
Comment (usmanatron) [reply devices/Tlc1543/Tlc1543.nfproj:45]: Yeah - I used the following Powershell:
`([system.reflection.assembly]::loadfile("filepath")).FullName`
This worked fine for all of them except this one assembly

**Comment by usmanatron** [comment]
Ok I've made those tiny changes.
One final question: Should I delete the `src/devices_generated/Tlc1543` folder?

**Comment by josesimoes** [comment]
> Ok I've made those tiny changes. One final question: Should I delete the `src/devices_generated/Tlc1543` folder?

@usmanatron yes please, so that's made clear the migration was performed.

**Comment by usmanatron** [comment]
Thanks @josesimoes - done!

**Comment by usmanatron** [comment]
@josesimoes - For avoidance of doubt I believe this is done from my side.  Let me know if you're waiting for anything else from me!

**Review by josesimoes — APPROVED** [review]
LGTM!

**Comment by josesimoes** [comment]
> @josesimoes - For avoidance of doubt I believe this is done from my side. Let me know if you're waiting for anything else from me!

@usmanatron all good! Just waiting if  @Ellerbach is OK with the latest changes.

**Comment by Ellerbach** [comment]
Thanks @usmanatron for the port!

**Pull request merged**